doc/vlan_stacking/vlan_stacking_HLD.md

```diff
@@ -0,0 +1,862 @@
+
+## Configuration
+
+* It is able to configure VLAN stacking by modifying CONFIG_DB or reload configuration by "config load" CLI command.
+* Invalid configuration with **_will_** be ignored and written an error in syslog. E.g., port is not exist.
+
+# High-Level Design
+
+## VlanMgr
+
+VlanMgr is used for check user configuration by DB schema, the VlanStack **_will_** parsing user configuration to SAI attributes and call SAI API.
+
+![VLAN Stacking create flow](./images/VlanStack%20config%20flow.png)
+
+VlanStack defined in Orch agent is used for all VLAN stacking related operations and keep the user configurations.
+
+* addEntry is used for add a C-VLAN ↔ S-VLAN mapping on the specified interface by user configuration. It function is used for updating the exist mapped, too. The update operation is implemented by remove old and then add new one, so it **_will_** effect the forwarding traffic when an entry be updating.
+* removeEntry is used to remove a C-VLAN ↔ S-VLAN mapping on the specified interface.
+
+```c++
+struct VlanStack
+{
+    std::map<VlanStackKey, VlanStackEntry> vlan_stack_map;
+
+    void addEntry(const VlanStackCfg& cfg);
+    void deleteEntry(const VlanStackKey& entry_key);
+};
+```
+
+## DB
+
+### CONFIG_DB
+
+A new table `VLAN_STACKING` is added in Config DB.
+
+```
+VLAN_STACKING|{{interface_name}}|{{s_vlanid}}
+"c_vlanids":        {{c_vlanids}}
+"s_vlan_priority":  {{s_vlan_priority}}
+```
+
+```
+`interface_name`  Ethernet port and channel port name
+`s_vlanid`        Service VLAN ID
+                  push VLAN for ingress direction
+                  pop (remove) VLAN for egress direction
+`c_vlanids`       match VLAN IDs on ingress direction of interface
+`s_vlan_priority` VLAN priority of Service VLAN
+```
+
+DB schema of VLAN_STACKING table:
+
+```
+; Defines schema for VLAN_STACKING configuration attributes
+key               = VLAN_STACKING:interface_name:s_vlanid
+interface_name    =
+s_vlanid          = vlan_id                               ; a number between 1 and 4094
+; field           = value
+c_vlanids         = vlan_id-or-range[,vlan_id-or-range]*  ; list of VLAN IDs
+s_vlan_priority   = %x30-37                               ; a number between 0 and 7
+
+; value annotations
+vlan_id           = %x31-39                               ; 1-9
+                  / %x31-39 DIGIT                         ; 10-99
+                  / %x31-39 2DIGIT                        ; 100-999
+                  / "1" 3DIGIT                            ; 1000-1999
+                  / "2" 3DIGIT                            ; 2000-2999
+                  / "3" 3DIGIT                            ; 3000-3999
+                  / "40" DIGIT %x30-34                    ; 4000-4094
+range             = vlan_id..vlan_id
+```
+
+A new table `VLAN_TRANSLATION` is added in Config DB.
+
+```
+VLAN_TRANSLATION|{{interface_name}}|{{s_vlanid}}
```

Comment:
Can we include Direction as the key in both VLAN_TRANSLATION & VALN_STACKING tables ? SAI proposal has the support for this via the attribute SAI_VLAN_STACK_ATTR_STAGE.
Some of the vendors support the partial translation rules, where the rule applies for Ingress direction only or  for Egress  direction only.